server/src/main/java/org/elasticsearch/search/profile/query/ProfileScorer.java

```diff
@@ -46,6 +48,20 @@
         nextDocTimer = profile.getTimer(QueryTimingType.NEXT_DOC);
         advanceTimer = profile.getTimer(QueryTimingType.ADVANCE);
         matchTimer = profile.getTimer(QueryTimingType.MATCH);
+        ProfileScorer profileScorer = null;
+        if (w.getQuery() instanceof ConstantScoreQuery && scorer instanceof ProfileScorer) {
+            profileScorer = (ProfileScorer) scorer;
+        } else if (w.getQuery() instanceof ConstantScoreQuery && scorer.getChildren().size() == 1) {
+            profileScorer = (ProfileScorer) scorer.getChildren().iterator().next().child;
+        }
```

Comment:
I understand when the `if` case may happen but I'm less sure about the second `else if`, when does it happen? Can you leave comments explaining?

Reply:
Yes, here is where the complexity is:

1) The first `if` is hit when you have a totalHits query that is not cached. (Test testConstantScoreTotalHitsQuery and cache ==false).

2) The `else if` is hit when you have a top N query. (Test testConstantScoreQuery)

3) There is a third case where we have a totalHits query and it is cached. In this case the provided scorer is not an instance of ProfileScorer and has no children. In that case we do not do any special treatment.